src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1133,6 +1135,19 @@ where
                         batch.source_upper,
                     );
 
+                    // This should always be true:
+                    // we only advance the `global_source_upper` PAST
+                    // a batch when that batch is fully processed. Offsets are required
+                    // (and in fact, we assert this above) to increase or stay the same
+                    // as we produce batches and a batch upper is always exactly 1
+                    // greater than the maximum offset in that batch. These
+                    // combined mean that the upper of one batch can never be greater
+                    // than the lower of the next.
+                    //
+                    // Our use of `source_upper_updates` ensures the `global_source_upper`
```

Comment:
It's not necessarily the same order, but we only process _source_upper updates_ once the frontier has advanced, ensuring that we have seen all batches that would have had a lower frontier.